### PR TITLE
Add exporting of Envoy metrics to HCP for linked clusters

### PR DESCRIPTION
**Pull request by jjti**
## Overview

This builds off the work of the Consul PR: https://github.com/hashicorp/consul/pull/20257

We want to let users send Envoy metrics to HCP for use in service-level dashboards and coming service topology map features. We also want to support that without requiring that users deploy a dedicated [Consul Telemetry Collector](https://developer.hashicorp.com/hcp/docs/consul/monitor/consul-central/observability/telemetry-collector). 

The behavior added to Consul Dataplane in this PR is:
1. Watch for changes to TelemetryState in Consul via the resource API
2. On creation of a TelemetryState resource, create an HTTP client to push metrics to HCP's Consul metrics endpoint
3. Start periodically (minutely) scraping Envoy's stats endpoint
4. Convert those Prometheus metrics to OTLP metrics
5. POST the metrics to HCP

Screenshot 2024-01-26 at 11 21 03 AM

More details are available in the RFC: https://docs.google.com/document/d/1gh3_zUtdIDeDPIH7z3d0-kkx1scB2oZLi7yuYg88PK8

### Unit/integration Testing

```
$go test -timeout 30s -tags integration github.com/hashicorp/consul-dataplane/pkg/hcp/telemetry

ok  	github.com/hashicorp/consul-dataplane/pkg/hcp/telemetry	0.788s	coverage: 90.7% of statements
```

```
go test -timeout 30s -tags integration github.com/hashicorp/consul-dataplane/pkg/hcp/telemetry/otlphttp

ok  	github.com/hashicorp/consul-dataplane/pkg/hcp/telemetry/otlphttp	0.651s	coverage: 88.9% of statements
```

### E2E Testing with HCP

To test this with real values I:
1. created a resource in consul-core by calling the resource API with grpcurl (using a real endpoint, filters, labels, and service principal)
2. validated that metrics were pushed through the HCP metrics endpoint and available for querying

```bash
# create a resource manually
grpcurl -d @ \
  -plaintext \
  -protoset pkg/consul.protoset \
  127.0.0.1:8502 \
  hashicorp.consul.resource.ResourceService.Write \
<<EOF
  {
    "resource": {
      "id": {
        "type": {
          "group": "hcp",
          "group_version": "v2",
          "kind": "TelemetryState"
        },
        "name": "global"
      },
      "metadata": {
        "consul.io/hcp/telemetry-state/debug/skip-deletion": "true"
      },
      "data": {
        "@type": "types.googleapis.com/hashicorp.consul.hcp.v2.TelemetryState",
        "resource_id": "organization/de9edc98-06f2-45b3-96ba-e4b0ced62717/project/a2c4c2d9-0087-4a1b-a9d6-b6030d55ea5c/hashicorp.consul.global-network-manager.cluster/josh-test-cdp-metrics",
        "client_id": "foo",
        "client_secret": "barr",
        "hcp_config": {
          "auth_url": "https://auth.idp.hcp.dev"
        },
        "proxy": {
          "http_proxy": "http://192.168.0.135:8080"
        },
        "metrics": {
          "endpoint": "http://consul-telemetry.hcp.dev/otlp/v1/metrics",
          "include_list": [".+"],
          "disabled": false,
          "labels": {
            "foo": "bar"
          }
        }
      }
    }
  }
EOF

```

Screenshot 2023-12-22 at 12 23 20 PM

### E2E Testing with HCP and a Proxy Server

`TelemetryState` has an HTTP proxy for use in environments where egress to the internet (ie HCP) is restricted to a single point of egress. This is similar to the [Datadog agent's proxy settings](https://docs.datadoghq.com/agent/configuration/proxy/?tab=linux).

```proto
message TelemetryState {
  ProxyConfig proxy = 5;
}

message ProxyConfig {
  // HttpProxy configures the http proxy to use for HTTP (non-TLS) requests.
  string http_proxy = 1;

  // HttpsProxy configures the http proxy to use for HTTPS (TLS) requests.
  string https_proxy = 2;

  // NoProxy can be configured to include domains which should NOT be forwarded through the configured http proxy
  repeated string no_proxy = 3;
}
```

To test pushing of metrics through a proxy, I set up Nginx locally as a forward proxy:

```
http {
...
    server {
        listen               8080;
        server_name  192.168.0.135;

        location /otlp/v1/metrics {
            proxy_ssl_verify off;
            proxy_pass https://consul-telemetry.hcp.dev/otlp/v1/metrics;
        }
```

And confirmed that metrics were pushed thru it to CTGW (both in logs in metrics in Grafana):

```
192.168.0.135 - - [25/Jan/2024:19:56:51 -0500] "POST http://consul-telemetry.hcp.dev/otlp/v1/metrics HTTP/1.1" 200 23 "-" "consul-dataplane/1.4.0-dev (linux/arm64)" "-"
```

### CODE REVIEW

File: pkg/dns/dns_test.go [hunk 1]
```diff
@@ -40,7 +46,7 @@ func genRandomBytes(size int) (blk []byte) {
 }
 
 func (s *DNSTestSuite) Test_DisabledServer() {
-	mockedDNSConsulClient := pbdns.NewMockDNSServiceClient(s.T())
+	mockedDNSConsulClient := NewMockDNSServiceClient(s.T())
```
Comment (jjti) [inline pkg/dns/dns_test.go:49]: the mock used in these tests was removed from `proto-public`

File: go.mod [hunk 2]
```diff
@@ -2,52 +2,77 @@ module github.com/hashicorp/consul-dataplane
...
-replace github.com/hashicorp/consul/proto-public v0.4.1 => github.com/hashicorp/consul/proto-public v0.1.2-0.20230929231147-632fd65c091c
-
 require (
 	dario.cat/mergo v1.0.0
 	github.com/armon/go-metrics v0.4.1
 	github.com/hashi-derek/grpc-proxy v0.0.0-20231207191910-191266484d75
 	github.com/hashicorp/consul-server-connection-manager v0.1.6
-	github.com/hashicorp/consul/proto-public v0.4.1
-	github.com/hashicorp/go-hclog v1.2.2
+	github.com/hashicorp/consul/proto-public v0.1.2-0.20240122214449-ea12838e3dc1
```
Comment (jjti) [inline go.mod:10]: TODO: replace with actual `consul/proto-public` version after release of: https://github.com/hashicorp/consul/pull/20257

File: pkg/hcp/telemetry/exporter_test.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+			expectScrape: true,
+			expectExport: true,
+		},
+		"disabled": {
+			modState: func(s *state) {
+				s.disabled = true
+			},
+		},
+		"disabled nil state": {
+			nilState: true,
```
Comment (johnbuonassisi) [inline pkg/hcp/telemetry/exporter_test.go:89]: Could you use `modeState` for this?

File: pkg/hcp/telemetry/exporter_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	} {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			r := require.New(t)
+
+			// Create a worker. We don't use the client or envoy admin addr for anything.
+			proxyID, err := uuid.GenerateUUID()
+			r.NoError(err)
+			worker := NewHCPExporter(NewMockResourceServiceClient(t), hclog.NewNullLogger(), "localhost:1234", proxyID)
```
Comment (johnbuonassisi) [inline pkg/hcp/telemetry/exporter_test.go:109]: nit: might be clearer to call this `exporter`

File: pkg/hcp/telemetry/metrics.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+func newScraper(envoyAdminHostPort string, logger hclog.Logger) scraper {
+	return &scraperImpl{
+		client:             http.DefaultTransport,
+		envoyAdminHostPort: envoyAdminHostPort,
+		logger:             logger,
+		parser:             new(expfmt.TextParser),
+	}
+}
+
+// scrape is called by the worker goroutine to scrape metrics from the Envoy admin API.
```
Comment (johnbuonassisi) [inline pkg/hcp/telemetry/metrics.go:47]:
```suggestion
// scrape is called by the exporter goroutine to scrape metrics from the Envoy admin API.
```

File: pkg/hcp/telemetry/state_test.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+/*
+The mocks in the adjacent files were created with:
+
+mockery --srcpkg=github.com/hashicorp/consul/proto-public/pbresource --name=ResourceServiceClient --output ./pkg/hcp/telemetry --outpkg telemetry --case underscore --testonly --structname MockResourceServiceClient
+mockery --srcpkg=github.com/hashicorp/consul/proto-public/pbresource --name=ResourceService_WatchListClient --output ./pkg/hcp/telemetry --outpkg telemetry --case underscore --testonly --structname MockResourceService_WatchListClient
+*/
```
Comment (johnbuonassisi) [inline pkg/hcp/telemetry/state_test.go:22]: Are these files missing?

File: pkg/hcp/telemetry/exporter.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"github.com/hashicorp/go-hclog"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+const exporterDefaultScrapeInterval = time.Minute
```
Comment (johnbuonassisi) [inline pkg/hcp/telemetry/exporter.go:12]: would it be useful to make this configurable?

File: pkg/hcp/telemetry/exporter_test.go (hunk 3 above)
Comment (johnbuonassisi) [inline pkg/hcp/telemetry/exporter_test.go:89]: if you wanted to re-use modState for this case, you could make modState return a *state.

**Review by johnbuonassisi — APPROVED** [review]
LGTM! Great work in here.

File: pkg/hcp/telemetry/metrics.go (hunk 5 above)
Comment (jjti) [reply pkg/hcp/telemetry/metrics.go]: agreed and makes sense. This and other references to workers were before a refactor/rename from worker > exporter. I just searched around for these remaining references.

File: pkg/hcp/telemetry/exporter.go (hunk 7 above)
Comment (jjti) [reply pkg/hcp/telemetry/exporter.go:12]: Ever? Definitely. We had a plan for something like that from Consul servers -- CTGW would tell servers how frequently to push -- but in practice we think 1m is a pretty good default (none of our dashboards/graphs use anything at a lower resolution than that right now). If we supported it already in CTGW I'd be more inclined to support it here

File: pkg/consuldp/consul_dataplane.go [hunk 8]
```diff
@@ -216,6 +221,14 @@ func (cdp *ConsulDataplane) Run(ctx context.Context) error {
 		return err
 	}
 
+	cdp.hcpTelemetry = telemetry.NewHCPExporter(
```
Comment (nickethier) [inline pkg/consuldp/consul_dataplane.go:224]: It doesn't look like we use this pointer anywhere else. Any reason not to do this?
```go
go telemetry.NewHCPExporter(...).Run(ctx)
```

**Review by nickethier — APPROVED** [review]

File: pkg/consuldp/consul_dataplane.go (hunk 8 above)
Comment (jjti) [reply pkg/consuldp/consul_dataplane.go]: That makes sense to me, just switched. I debated just changing the signature to `go telemetry.Run(ctx, ...params)`, might still do that...

File: pkg/dns/dns_test.go [hunk 9]
```diff
@@ -40,7 +41,7 @@ func genRandomBytes(size int) (blk []byte) {
 }
 
 func (s *DNSTestSuite) Test_DisabledServer() {
-	mockedDNSConsulClient := pbdns.NewMockDNSServiceClient(s.T())
```
Comment (jjti) [inline pkg/dns/dns_test.go:43]: `NewMockDNSServiceClient` was removed from consul/proto-public, this moves the mock reference to the one generated in this repo

**Review by david-yu — APPROVED** [review]
Approved we'd like to get this into 1.18 GA

File: .changelog/406.txt [hunk 10]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
```
Comment (jjti) [inline .changelog/406.txt:1]: TODO: remove this

File: pkg/hcp/telemetry/metrics.go [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+)
+
+var (
+	errScraperFailedToCreateRequest = errors.New("failed to create request")
+	errScraperFailedToIssueRequest  = errors.New("failed to issue request")
+	errScraperFailedStatusCode      = errors.New("non-200 status code")
+	errScraperFailedToParseResponse = errors.New("failed to parse response")
+)
+
+type scraper interface {
```
Comment (jjti) [inline pkg/hcp/telemetry/metrics.go:27]: TODO: rename this file to scraper

File: cmd/consul-dataplane/main.go [hunk 12]
```diff
@@ -184,6 +189,15 @@ func run() error {
...
 	}()
 
+	go func() {
+		// Configure debug and runtime metrics endpoints.
+		// TODO: this is just for testing a PR - remove.
+		http.Handle("/debug/metrics", promhttp.Handler())
+		if err := http.ListenAndServe(":2112", nil); err != nil {
+			hclog.Default().Error("failed to start debug metrics server", "error", err)
+		}
+	}()
```
Comment (jjti) [inline cmd/consul-dataplane/main.go:199]: TODO: remove before merging

**Pull request merged**